File: LeetCode/c++/RomanToInteger.cpp
#include <iostream>
#include <string>
#include <map>

using namespace std; 

//建立罗马数字和阿拉伯数字的映射
int romanToInt(string s) 
{
	map<char, int> m;
	m['I'] = 1;
	m['V'] = 5;
	m['X'] = 10;
	m['L'] = 50;
	m['C'] = 100;
	m['D'] = 500;
	m['M'] = 1000;
	
	//如果值小的数字在右边表示加；在左边表示减
	int result = 0;
	for (int i = 0; i < s.size() - 1; i++)
	{
		if (m[s[i]] >= m[s[i + 1]]) //大于等于比右边的数，用加
			result += m[s[i]];
		else
			result -= m[s[i]];
	}

	//处理s[s.size()-1]的情况,结尾的数字肯定是加上去的
	result += m[s[s.size() - 1]];

	return result;
}

int main()
{

	return 0;
}

File: LeetCode/iter1/python/RomanToInteger.py
class Solution:
    # @param {string} s
    # @return {integer}
    def romanToInt(self, s):
        #建立罗马数字与阿拉伯数的映射
        romanDict = {'I':1,'V':5,'X':10,'L':50,'C':100,'D':500,'M':1000}
        sSize = len(s)
        result=0
        for i in range(sSize-1):
            #如果前一个数大于等于后一个数就加1
            front = romanDict[s[i]]
            latter = romanDict[s[i+1]]			
            if(front>=latter):
                result += front 
            else:
                result -= front			
				
        result += romanDict[s[sSize-1]]
		
        return result

#测试程序
s=Solution()
print(s.romanToInt("I"))
print(s.romanToInt("III"))
print(s.romanToInt("IIII"))
print(s.romanToInt("XIX"))
print(s.romanToInt("XIV"))
print(s.romanToInt("XXVII"))